container/health.go

```diff
@@ -0,0 +1,63 @@
+package container
+
+import (
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/engine-api/types"
+)
+
+// States
+const (
+	// Health probe has not yet run successfully and all completed runs
+	// started within the grace period.
+	Starting = "starting"
+
+	// Last probe was successful.
+	Healthy = "healthy"
+
+	// Last probe started after the grace period and was unsuccessful
+	// (failed or exceeded the timeout).
+	Unhealthy = "unhealthy"
+)
+
+// Health holds the current container health-check state
+type Health struct {
+	types.Health
+	stop chan struct{} // Write struct{} to stop the monitor
+}
+
+// String returns a human-readable description of the health-check state
+func (s *Health) String() string {
+	if s.stop == nil {
+		return "no healthcheck"
+	}
+	switch s.Status {
+	case Starting:
+		return "health: starting"
+	default: // Healthy and Unhealthy are clear on their own
+		return s.Status
+	}
+}
+
+// OpenMonitorChannel creates and returns a new monitor channel. If there already is one,
+// it returns nil.
+func (s *Health) OpenMonitorChannel() chan struct{} {
+	if s.stop == nil {
```

Comment:
Opening and closing the channel only happens from the event monitor thread, which already runs with the container locked.